### PR TITLE
Complete scene return for component = 0 (#1067)

### PR DESCRIPTION
**Pull request by j-petit**
### Description
When calling the /get_planning_scene service with component = 0, the service will return the full scene instead of an empty one (proposed in #1067)
I identified a single place in the codebase where a call to the service with all components takes place and replaced it with a call with components = 0. I am not to happy about writing it out as it is (that's why I had to add the comment), maybe someone can give me a hint to make it more understandable intuitively.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Document API changes relevant to the user in the moveit/MIGRATION.md notes
- [ ] Decide if this should be cherry-picked to other current ROS branches

### CODE REVIEW

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp [hunk 1]
```diff
@@ -471,12 +471,8 @@ bool PlanningSceneMonitor::requestPlanningSceneState(const std::string& service_
...
   ros::ServiceClient client = ros::NodeHandle().serviceClient<moveit_msgs::GetPlanningScene>(service_name);
   moveit_msgs::GetPlanningScene srv;
-  srv.request.components.components =
-      srv.request.components.SCENE_SETTINGS | srv.request.components.ROBOT_STATE |
-      srv.request.components.ROBOT_STATE_ATTACHED_OBJECTS | srv.request.components.WORLD_OBJECT_NAMES |
-      srv.request.components.WORLD_OBJECT_GEOMETRY | srv.request.components.OCTOMAP |
-      srv.request.components.TRANSFORMS | srv.request.components.ALLOWED_COLLISION_MATRIX |
-      srv.request.components.LINK_PADDING_AND_SCALING | srv.request.components.OBJECT_COLORS;
+  // Requests all components if 0 is specified
+  srv.request.components.components = 0;
```
Comment (felixvd) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: This is already initialized to 0. Would remove this line and reword the comment to "All scene components are returned if none are specified" or similar.

File: moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp [hunk 2]
```diff
@@ -54,7 +54,26 @@ bool move_group::MoveGroupGetPlanningSceneService::getPlanningSceneService(movei
   if (req.components.components & moveit_msgs::PlanningSceneComponents::TRANSFORMS)
     context_->planning_scene_monitor_->updateFrameTransforms();
   planning_scene_monitor::LockedPlanningSceneRO ps(context_->planning_scene_monitor_);
-  ps->getPlanningSceneMsg(res.scene, req.components);
+
+  if (!req.components.components)
```
Comment (felixvd) [inline moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp]: The same comment would be helpful here: "Return all scene components if none are specified"

File: moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp [hunk 3]
```diff
@@ -54,7 +54,26 @@ bool move_group::MoveGroupGetPlanningSceneService::getPlanningSceneService(movei
...
+  {
+    moveit_msgs::PlanningSceneComponents all_components;
+    all_components.components =
```
Comment (felixvd) [inline moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp]: `INT_MAX`?

**Comment by j-petit** [comment]
I searched through the codebase thoroughly and am quite confident that there are no calls to the service where an empty scene response is expected. The service is called only through an intermediate function which always requests the full scene. Nevertheless, I would still like to do the performance check but I don't know what exactly do you mean when referring to performance and how would you measure it. Could you give me a hint how to do this?

**Comment by felixvd** [comment]
The keyword is code benchmarking, and I would simply test if calling the `get_planning_scene` service is slower with `req.components.components = 0` and `req.components.components = UINT_MAX`. If you call the service from a Python script a bunch of times and time the execution, that should be enough. There are much better tools for this, but I don't know them. This might be overkill anyway; someone with more compiler knowledge would know if this will affect performance.

Apart from that, I also had a look through the code and couldn't find any calls to this service except the `PlanningSceneMonitor::requestPlanningSceneState` wrapper which you changed. I think the change itself is fine.

To document it, I would personally add the same comment to the [PlanningSceneRequest msg file](https://github.com/ros-planning/moveit_msgs/blob/melodic-devel/msg/PlanningSceneComponents.msg) and maybe a line like "get_planning_scene service now returns the whole scene if no components are specified" to the migration notes. Maybe some maintainers can give their opinion and request those changes. Apart from that and the first paragraph, I think this is fine.

That said, I haven't tested it.

**Comment by j-petit** [comment]
> The keyword is code benchmarking, and I would simply test if calling the `get_planning_scene` service is slower with `req.components.components = 0` and `req.components.components = UINT_MAX`. If you call the service from a Python script a bunch of times and time the execution, that should be enough. There are much better tools for this, but I don't know them. This might be overkill anyway; someone with more compiler knowledge would know if this will affect performance.
> 
I did some measurements and the difference is tiny. So I think we are good to go from this aspect.

> To document it, I would personally add the same comment to the [PlanningSceneRequest msg file](https://github.com/ros-planning/moveit_msgs/blob/melodic-devel/msg/PlanningSceneComponents.msg) and maybe a line like "get_planning_scene service now returns the whole scene if no components are specified" to the migration notes. Maybe some maintainers can give their opinion and request those changes. Apart from that and the first paragraph, I think this is fine.
>
I think that the comment is better fit in the service description (moveit_msgs/srv/GetPlanningScene.srv) as this is a property of the service and not the message itself?

**Comment by felixvd** [comment]
Yes, the [service description file](https://github.com/ros-planning/moveit_msgs/blob/melodic-devel/srv/GetPlanningScene.srv) is the right place. I got confused copying things around.

File: MIGRATION.md [hunk 4]
```diff
@@ -2,6 +2,12 @@
 
 API changes in MoveIt! releases
 
+## ROS Noetic (upcoming changes in master)
+- Extended the return value of `MoveitCommander.MoveGroup.plan()` from `trajectory` to a tuple of `(success, trajectory, planning_time, error_code)` to better match the C++ MoveGroupInterface ([790](https://github.com/ros-planning/moveit/pull/790/))
+- `moveit_rviz.launch`, generated by MSA, provides an argument `rviz_config` to configure the rviz config to be used. The old boolean config argument was dropped. ([1397](https://github.com/ros-planning/moveit/pull/1397))
+- Moved the example package `moveit_controller_manager_example` into [moveit_tutorials](https://github.com/ros-planning/moveit_tutorials)
+- Requests to `get_planning_scene` service without explicitely setting components returns full scene
```
Comment (felixvd) [inline MIGRATION.md]: explicitly setting "components" now return the full scene

**Comment by j-petit** [comment]
Sorry, when trying to get a clean PR through incorporating changes from master to my devel-branch, I messed it up. Cleaning up now...

File: moveit_core/robot_state/src/robot_state.cpp [hunk 5]
```diff
@@ -999,55 +985,55 @@ bool RobotState::clearAttachedBody(const std::string& id)
     return false;
 }
 
-const Eigen::Isometry3d& RobotState::getFrameTransform(const std::string& id)
+const Eigen::Isometry3d& RobotState::getFrameTransform(const std::string& frame_id)
```
Comment (felixvd) [inline moveit_core/robot_state/src/robot_state.cpp]: These changes should clear up if you rebase your branch on master.

Comment (j-petit) [reply moveit_core/robot_state/src/robot_state.cpp]: Thanks, worked.

File: MIGRATION.md [hunk 6]
```diff
@@ -6,6 +6,7 @@ API changes in MoveIt! releases
 - Extended the return value of `MoveitCommander.MoveGroup.plan()` from `trajectory` to a tuple of `(success, trajectory, planning_time, error_code)` to better match the C++ MoveGroupInterface ([790](https://github.com/ros-planning/moveit/pull/790/))
 - `moveit_rviz.launch`, generated by MSA, provides an argument `rviz_config` to configure the rviz config to be used. The old boolean config argument was dropped. ([1397](https://github.com/ros-planning/moveit/pull/1397))
 - Moved the example package `moveit_controller_manager_example` into [moveit_tutorials](https://github.com/ros-planning/moveit_tutorials)
+- Requests to `get_planning_scene` service without explicitely setting "components" now returns full scene
```
Comment (felixvd) [inline MIGRATION.md]: -> explicitly

Comment (felixvd) [reply MIGRATION.md]: -> return

Comment (j-petit) [reply MIGRATION.md]: Sorry, missed my spelling mistakes.

File: MIGRATION.md [hunk 7]
```diff
@@ -6,6 +6,7 @@ API changes in MoveIt! releases
 - Extended the return value of `MoveitCommander.MoveGroup.plan()` from `trajectory` to a tuple of `(success, trajectory, planning_time, error_code)` to better match the C++ MoveGroupInterface ([790](https://github.com/ros-planning/moveit/pull/790/))
 - `moveit_rviz.launch`, generated by MSA, provides an argument `rviz_config` to configure the rviz config to be used. The old boolean config argument was dropped. ([1397](https://github.com/ros-planning/moveit/pull/1397))
 - Moved the example package `moveit_controller_manager_example` into [moveit_tutorials](https://github.com/ros-planning/moveit_tutorials)
+- Requests to `get_planning_scene` service without explicitly setting "components" now return full scene
```
Comment (rhaschke) [inline MIGRATION.md]:
```suggestion
- Requests to `get_planning_scene` service without explicitly setting "components" now returns full scene
```

File: moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp [hunk 8]
```diff
@@ -54,7 +54,19 @@ bool move_group::MoveGroupGetPlanningSceneService::getPlanningSceneService(movei
   if (req.components.components & moveit_msgs::PlanningSceneComponents::TRANSFORMS)
     context_->planning_scene_monitor_->updateFrameTransforms();
   planning_scene_monitor::LockedPlanningSceneRO ps(context_->planning_scene_monitor_);
-  ps->getPlanningSceneMsg(res.scene, req.components);
+
+  // all scene components are returned if none are specified
+  if (!req.components.components)
```
Comment (rhaschke) [inline moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp]: I suggested a more compact code snippet.

File: MIGRATION.md (hunk 7 above)
Comment (j-petit) [reply MIGRATION.md]: No, I think this is correct, as requests is plural.

File: moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp (hunk 8 above)
Comment (j-petit) [reply moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp]: thanks!

File: moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp [hunk 9]
```diff
@@ -55,17 +55,8 @@ bool move_group::MoveGroupGetPlanningSceneService::getPlanningSceneService(movei
...
-    moveit_msgs::PlanningSceneComponents all_components;
-    all_components.components = UINT_MAX;
-    ps->getPlanningSceneMsg(res.scene, all_components);
-  }
-  else
-  {
-    ps->getPlanningSceneMsg(res.scene, req.components);
-  }
+  static const moveit_msgs::PlanningSceneComponents all_components{ UINT_MAX };
+  ps->getPlanningSceneMsg(res.scene, req.components.components ? req.components : all_components);
```
Comment (felixvd) [inline moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp]: Pretty, but I'd keep the comment `// Return all scene components if none are specified`.

Comment (rhaschke) [reply moveit_ros/move_group/src/default_capabilities/get_planning_scene_service_capability.cpp]: Fair enough!

File: MIGRATION.md (hunk 7 above)
Comment (rhaschke) [reply MIGRATION.md]: You are right.

**Comment by BryceStevenWilley** [comment]
Travis failed because it looks like `rosdep` (or maybe github) was down a few days ago. I'm going to close and re-open this PR to retrigger the build.

**Comment by BryceStevenWilley** [comment]
Attempting to fix flaky CI [here](https://github.com/ros-planning/moveit_ci/pull/63).

**Comment by BryceStevenWilley** [comment]
Trying one more time to get Travis to pass.

**Pull request merged**